### ISSUE TITLE
TypeError: Cannot read properties of undefined (reading 'name')

### ISSUE THREAD

**Issue by nixmeer** [henry-spanka/freeathome-api#45]
Hello :) 

I got an issue using the homebridge-freeathome plugin. The error message is as follows:

```
[31/01/2022, 21:48:47] TypeError: Cannot read properties of undefined (reading 'name')
     at Function.parseDeviceData (/usr/local/lib/node_modules/homebridge-freeathome/node_modules/freeathome-api/dist/lib/XmlParser.js:110:106)
     at Function.parseMasterUpdate (/usr/local/lib/node_modules/homebridge-freeathome/node_modules/freeathome-api/dist/lib/XmlParser.js:17:21)
     at SystemAccessPoint. (/usr/local/lib/node_modules/homebridge-freeathome/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:314:53)
     at Generator.next ()
     at /usr/local/lib/node_modules/homebridge-freeathome/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:8:71
     at new Promise ()
     at __awaiter (/usr/local/lib/node_modules/homebridge-freeathome/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:4:12)
     at SystemAccessPoint.handleRpcCallResult (/usr/local/lib/node_modules/homebridge-freeathome/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:293:16)
     at SystemAccessPoint. (/usr/local/lib/node_modules/homebridge-freeathome/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:267:26)
     at Generator.next () [31/01/2022, 21:48:47] Got SIGTERM, shutting down Homebridge... [31/01/2022, 21:48:52] [HB Supervisor] Homebridge Process Ended. Code: 143, Signal: null
```

I'm not a javascript expert but digging a little into the `XmlParser.js` file I was able to get the  device variable. It contained the following:

```

<device softwareId="1" softwareVersion="2.6.4" deviceId="FFFF" nameId="02BE" sysAp="true" domainAddress="CB64" individualAddress="A096" profile="0810" maxAPDULength="37" compilerVersion="00000000" buildNumber="00000000" iconId="FFF8" protocolVersion="00FF" minConfigVersion="0000" deviceFlavor="00" functionId="FEFD" interface="sysap" batteryLow="false" readOnlyDisplayName="true" serialNumber="ABB700000000" commissioningState="ready">
    <attribute name="lastReset">20220215184558869</attribute>
    <channels>
        <channel mask="FFFFFFFF" nameId="01BF" sameLocation="true" i="ch0000" cid="ABB7FFF3">
            <attribute name="functionId">25</attribute>
            <inputs>
                <dataPoint i="idp0000">
                <value>0</value>
                <address timeSlot="0" primary="true">8125</address>
                </dataPoint>
            </inputs>
            <outputs>
                <dataPoint i="odp0000">
                    <value>0</value>
                    <address timeSlot="0" primary="true">a2e0</address>
                </dataPoint>
            </outputs>
            <parameters/>
        </channel>
        <channel mask="FFFFFFFF" nameId="01C1" hidden="true" i="ch0001" cid="ABB7FFFB">
            <attribute name="functionId">26</attribute>
            <inputs/>
            <outputs>
                <dataPoint i="odp0000">
                    <value>13:38:30</value>
                    <address timeSlot="0" primary="true">7c66</address>
                </dataPoint>
                <dataPoint i="odp0001">
                    <value>2022-02-16</value>
                    <address timeSlot="0" primary="true">24ae</address>
                </dataPoint>
                <dataPoint i="odp0002">
                    <value>0</value>
                    <address timeSlot="0" primary="true">6f12</address>
                </dataPoint>
            </outputs>
            <parameters/>
        </channel>
        <channel mask="FFFFFFFF" nameId="04CE" hidden="true" i="ch0002" cid="ABB7FFF1">
            <attribute name="functionId">86</attribute>
            <inputs/>
            <outputs/>
            <parameters/>
        </channel>
    </channels>
</device>
```

Why does this error occur? Is the SysAp firmware to old?

**Comment by nixmeer** [comment]
Tested version SysAp Firmware 3.0.0. Same issue.

```
<device readOnlyDisplayName="true" softwareId="1" softwareVersion="3.0.0" nameId="02BE" deviceId="FFFF" sysAp="true" domainAddress="CB64" individualAddress="A029" profile="0810" maxAPDULength="37" compilerVersion="00000000" buildNumber="00000000" iconId="0020" protocolVersion="00FF" minConfigVersion="0000" deviceFlavor="00" functionId="0020" interface="sysap" serialNumber="ABB700000000" commissioningState="ready">
    <channels>
        <channel mask="FFFFFFFF" nameId="01BF" persistentId="0000" sameLocation="true" i="ch0000" cid="ABB7FFF3">
            <attribute name="functionId">25</attribute>
            <attribute name="selectedIcon">4</attribute>
            <inputs>
                <dataPoint i="idp0000">
                    <value>0</value>
                    <address timeSlot="0" primary="true">8125</address>
                </dataPoint>
            </inputs>
            <outputs>
                <dataPoint i="odp0000">
                    <value>0</value>
                    <address timeSlot="0" primary="true">a2e0</address>
                </dataPoint>
            </outputs>
            <parameters />
        </channel>
        <channel mask="FFFFFFFF" nameId="01C1" persistentId="0001" hidden="true" i="ch0001" cid="ABB7FFFB">
            <attribute name="functionId">26</attribute>
            <inputs />
            <outputs>
                <dataPoint i="odp0000">
                    <value>14:57:30</value>
                    <address timeSlot="0" primary="true">7c66</address>
                </dataPoint>
                <dataPoint i="odp0001">
                    <value>2022-02-16</value>
                    <address timeSlot="0" primary="true">24ae</address>
                </dataPoint>
                <dataPoint i="odp0002">
                    <value>0</value>
                    <address timeSlot="0" primary="true">6f12</address>
                </dataPoint>
            </outputs>
            <parameters />
        </channel>
        <channel mask="FFFFFFFF" nameId="04CE" persistentId="0002" hidden="true" i="ch0002" cid="ABB7FFF1">
            <attribute name="functionId">86</attribute>
            <inputs />
            <outputs />
            <parameters />
        </channel>
    </channels>
</device>
```

**Comment by nixmeer** [comment]
I downgraded to homebridge-freeathome v2.0.2. It didn't work. I then quit the homebridge project and started to add thermostats.
It suddenly started working. However, devices took forever to execute the commands and it stopped working after restarting homebridge.

I can provide a homebridge log file with debug output, if it helps.

**Comment by henry-spanka** [comment]
Make sure each device in your home has a name.
Try to login into the free@home web panel and turn on/off a few accessories. Then restart the API. This should hopefully resolve the problem.

**Comment by nixmeer** [comment]
Every device had a name and a room assigned to it. And neither turning on and off nor deleting and re-adding the devices helped. 

As you can see above, the error occurred when it tried to parse the xml of the SysAp which has no floor and no room assigned to it. I wonder if the SysAp is supposed to be parsed by `Function.parseDeviceData` since it has some kind of special status

**Comment by b3ss1k** [comment]
Hello, I wanted to ask such a question, for some reason I do not add curtains to homebirdge for some unknown reason. Moreover, I have several objects and everything works fine on some, and on some curtains are simply not visible. Any suggestions why it might not work ?

**Comment by bab200** [comment]
anyone resolved this? I had the API running - but now I got this same error

pi@raspberrypi:~ $ node node_modules/freeathome-api/bin/freeathome-api
22/01/2023, 19:31:44 - INFO: Starting free@home API
22/01/2023, 19:31:48 - INFO: Connected as installer@busch-jaeger.de/freeathome-api
22/01/2023, 19:31:48 - INFO: Websocket Server started
22/01/2023, 19:31:48 - INFO: Webserver Started
22/01/2023, 19:31:48 - INFO: Sent Authenticator
22/01/2023, 19:31:48 - INFO: Received Local Key
22/01/2023, 19:31:48 - INFO: Sent New Session Request
22/01/2023, 19:31:48 - INFO: Sent Login SASL Message
22/01/2023, 19:31:48 - INFO: Received SASL Challenge
22/01/2023, 19:31:48 - INFO: Sent SASL Challenge Response
22/01/2023, 19:31:48 - INFO: Received SASL Login Confirmation
22/01/2023, 19:31:48 - INFO: Successfully Authenticated
22/01/2023, 19:31:48 - INFO: Announced Capabilities to System Access Point
22/01/2023, 19:31:48 - INFO: Requested Master Data Structure from System Access Point
22/01/2023, 19:31:52 - INFO: Received Master Update from System Acccess Point
/home/pi/node_modules/freeathome-api/dist/appRunner.js:34
        throw up;
        ^

TypeError: Cannot read properties of undefined (reading 'name')
    at XmlParser.parseDeviceData (/home/pi/node_modules/freeathome-api/dist/lib/XmlParser.js:111:106)
    at XmlParser.parseMasterUpdate (/home/pi/node_modules/freeathome-api/dist/lib/XmlParser.js:18:21)
    at SystemAccessPoint.<anonymous> (/home/pi/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:326:53)
    at Generator.next (<anonymous>)
    at /home/pi/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/home/pi/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:4:12)
    at SystemAccessPoint.handleRpcCallResult (/home/pi/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:305:16)
    at SystemAccessPoint.<anonymous> (/home/pi/node_modules/freeathome-api/dist/lib/SystemAccessPoint.js:273:26)
    at Generator.next (<anonymous>)

Node.js v19.4.0

**Comment by henry-spanka** [comment]
Hello @bab200

Please enable debug logging and send me the log to my email address (see GitHub profile).

Set the key "debug" in the config file to "true" (see https://github.com/henry-spanka/freeathome-api/blob/master/config.example.json). 
e.g. `node node_modules/freeathome-api/bin/freeathome-api > debug.log` (run for 30 secs then Ctrl+C)

**Comment by henry-spanka** [comment]
Assigning a device to a floor but outside any room will result in a crash. The issue has now been fixed in *v1.1.3*. As a workaround you can also make sure that all devices are placed inside a room.
